File: backend/routers/notifications.py
from fastapi import APIRouter, Depends, HTTPException, Request, Response
from fastapi.concurrency import run_in_threadpool

from auth import get_current_user
from config import get_settings
from database import get_supabase
from models import NotificationPreferences, PushSubscriptionCreate, PushUnsubscribe
from rate_limit import limiter
from services.notification_copy import notification_text
from services.push_service import send_to_user

router = APIRouter(prefix="/notifications", tags=["notifications"])

_DEFAULT_PREFERENCES = NotificationPreferences().model_dump()


@router.post("/subscribe", status_code=204)
async def subscribe(payload: PushSubscriptionCreate, user=Depends(get_current_user)):
    """Registers (or re-registers) one device for this user.

    Two-step, so one physical device can never end up with more than one
    deliverable row (the root cause of the duplicate-notification bug):

    1. If the client sent a `device_id` (its stable per-install UUID), delete
       any OTHER row for this user carrying that same `device_id` — that row
       is a pre-rotation orphan whose `endpoint` the browser has already
       moved on from but which a push service may still deliver to for a
       while. This is what a plain endpoint-keyed upsert could never clean up
       (a new endpoint just looks like a brand-new row to it).
    2. Upsert on `endpoint` (globally unique): an unchanged-endpoint
       resubscribe — the common app-open resync — just refreshes keys and
       `last_seen_at` in place.

    Deliberately does NOT touch notification_preferences.push_enabled — the
    frontend flips that separately via PUT /preferences right after a
    successful subscribe, so "have I granted a subscription" and "does the
    user currently want pushes" stay two independent facts (a user can hold a
    live subscription while push is toggled off, and get it back on without a
    fresh permission prompt)."""
    supabase = get_supabase()
    row = {
        "user_id": user.id,
        "endpoint": payload.endpoint,
        "p256dh": payload.keys.p256dh,
        "auth_key": payload.keys.auth,
        "user_agent": payload.user_agent,
        "device_id": payload.device_id,
        "last_seen_at": "now()",
    }

    def _persist() -> None:
        if payload.device_id:
            supabase.table("push_subscriptions").delete().eq("user_id", user.id).eq(
                "device_id", payload.device_id
            ).neq("endpoint", payload.endpoint).execute()
        supabase.table("push_subscriptions").upsert(row, on_conflict="endpoint").execute()

    await run_in_threadpool(_persist)
    return None


@router.delete("/subscribe", status_code=204)
async def unsubscribe(payload: PushUnsubscribe, user=Depends(get_current_user)):
    """Best-effort: the frontend calls this both on an explicit "turn off
    notifications" action and opportunistically whenever it detects its own
    subscription is no longer valid. Scoped to `.eq("user_id", user.id)` on
    top of the endpoint match — endpoint is already globally unique, but this
    keeps the ownership-filter discipline CLAUDE.md requires on every query
    against the service-role client, rather than relying on uniqueness
    alone."""
    supabase = get_supabase()
    await run_in_threadpool(
        lambda: supabase.table("push_subscriptions").delete().eq("endpoint", payload.endpoint).eq("user_id", user.id).execute()
    )
    return None


@router.get("/preferences", response_model=NotificationPreferences)
async def get_preferences(user=Depends(get_current_user)):
    supabase = get_supabase()
    result = await run_in_threadpool(
        lambda: supabase.table("notification_preferences").select("*").eq("user_id", user.id).maybe_single().execute()
    )
    data = (result.data if result else None) or {}
    # Falls back to defaults rather than 404ing: sql/schema.sql's trigger
    # auto-creates this row for every new signup, but an account that
    # existed before this migration ran only gets one from the one-time
    # backfill statement in that same migration — if that hasn't been
    # (re-)run yet on this Supabase project, this keeps the endpoint working
    # instead of erroring for those users.
    return {**_DEFAULT_PREFERENCES, **data}


@router.put("/preferences", response_model=NotificationPreferences)
async def update_preferences(payload: NotificationPreferences, user=Depends(get_current_user)):
    supabase = get_supabase()
    row = {"user_id": user.id, **payload.model_dump(), "updated_at": "now()"}

    def _persist() -> None:
        supabase.table("notification_preferences").upsert(row, on_conflict="user_id").execute()
        # push_enabled is the single source of truth for "this account wants
        # pushes at all". The scheduler skips a push_enabled=false user
        # entirely, so every push_subscriptions row for them is dead weight
        # the moment this flips off — and a lingering, still-deliverable row
        # is exactly how a later re-enable ends up delivering duplicates.
        # Enforce "push off => zero subscriptions" here so it holds even when
        # the client's own browser-side unsubscribe + DELETE /subscribe
        # didn't land (flaky network, tab closed mid-toggle, a second device
        # that never heard about the toggle). Each device re-registers itself
        # on its next app open once push is back on.
        if payload.push_enabled is False:
            supabase.table("push_subscriptions").delete().eq("user_id", user.id).execute()

    await run_in_threadpool(_persist)
    return payload


@router.post("/test", status_code=200)
@limiter.limit("5/minute;2/10 seconds")
async def send_test_notification(request: Request, response: Response, language: str = "en", user=Depends(get_current_user)):
    """Settings → Notifications' "send yourself a test" button — the
    fastest way for a user to confirm push actually reaches their device
    without waiting for a real reminder to come due. Rate-limited tighter
    than this API's app-wide default since, unlike the CRUD routes above,
    this one has a real external side effect (an actual push service call)
    per request.

    `language` is a query param the frontend fills with its own current UI
    language (same `?language=` convention as GET /coach/weekly-recap) —
    simpler than a second DB read of notification_preferences.language just
    to answer "which language is this specific button click in", and always
    correct even for a language switch that hasn't been saved yet."""
    settings = get_settings()
    if not settings.vapid_configured:
        raise HTTPException(status_code=503, detail="Push notifications aren't configured on this server")
    lang = "ro" if language == "ro" else "en"
    title, body = notification_text(lang, "test")
    sent = await run_in_threadpool(
        lambda: send_to_user(user.id, {"title": title, "body": body, "url": "/", "tag": "test"})
    )
    if sent == 0:
        raise HTTPException(status_code=404, detail="No active push subscriptions found for this device/account")
    return {"sent": sent}

File: frontend/js/notifications.js
import { api } from "./api.js";
import { VAPID_PUBLIC_KEY } from "./config.js";
import { showToast } from "./ui.js";
import { getLanguage, onLanguageChange, t } from "./i18n.js";

// Real Web Push (VAPID) — replaces the old local-only, tab-must-be-open
// reminder system (frontend/js/reminders.js, removed). The firing decision
// now lives entirely server-side (backend/services/notification_scheduler.py)
// so this module's only job is: (1) keep the browser's push subscription and
// this device's copy of it on the backend in sync, and (2) sync the settings
// UI toggles against GET/PUT /notifications/preferences. Nothing here polls
// or fires a notification itself anymore — doing that from an open tab
// alongside the server would risk a duplicate for the same event.
//
// Defaults deliberately mirror sql/schema.sql's own column defaults exactly
// (push_enabled false — a subscription still requires a real user gesture —
// everything else true/on) so that the very first time a user flips the
// master toggle, they land on a fully-configured, motivating setup with zero
// extra taps: a 7 PM daily reminder, smart nudges, and the weekly recap —
// see reminders.defaultsHint in index.html for the copy that tells them so.
const DEFAULT_PREFERENCES = {
  push_enabled: false,
  daily_reminder_enabled: true,
  reminder_mode: "fixed",
  daily_reminder_time: "19:00",
  reminder_interval_hours: 4,
  smart_nudges_enabled: true,
  weekly_recap_enabled: true,
  quiet_hours_start: "22:00",
  quiet_hours_end: "08:00",
  language: "en",
};

let preferences = { ...DEFAULT_PREFERENCES };
// Guards the language-change listener below: don't PUT a language-only patch
// before the real GET /notifications/preferences has ever resolved, or it
// would overwrite every other field with DEFAULT_PREFERENCES' placeholders.
let preferencesLoaded = false;

const el = (id) => document.getElementById(id);

function pushSupported() {
  return "serviceWorker" in navigator && "PushManager" in window && "Notification" in window;
}

function urlBase64ToUint8Array(base64String) {
  const padding = "=".repeat((4 - (base64String.length % 4)) % 4);
  const base64 = (base64String + padding).replace(/-/g, "+").replace(/_/g, "/");
  const rawData = atob(base64);
  return Uint8Array.from(rawData, (char) => char.charCodeAt(0));
}

// A stable per-install id, generated once and kept forever. This — not the
// push `endpoint`, which the browser silently rotates on its own (key
// refresh, storage pressure) — is what the backend upserts on to guarantee
// exactly ONE subscription row per device: every rotation lands back on the
// same row instead of leaving the pre-rotation endpoint behind as a
// still-briefly-deliverable orphan (the duplicate-notification bug). See
// sql/schema.sql's push_subscriptions.device_id comment.
const DEVICE_ID_KEY = "ironlog:device-id";
let sessionDeviceId = null;
function randomId() {
  try {
    if (crypto?.randomUUID) return crypto.randomUUID();
  } catch {
    /* fall through */
  }
  return `dev-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 12)}`;
}
function getDeviceId() {
  try {
    let id = localStorage.getItem(DEVICE_ID_KEY);
    if (!id) {
      id = randomId();
      localStorage.setItem(DEVICE_ID_KEY, id);
    }
    return id;
  } catch {
    // localStorage blocked (private mode / partitioned iframe): a
    // per-session id still collapses THIS session's rotations, just not
    // across restarts.
    if (!sessionDeviceId) sessionDeviceId = randomId();
    return sessionDeviceId;
  }
}

// True iff `subscription` was minted with exactly `vapidKey` as its
// applicationServerKey. A subscription left over from a different (old) VAPID
// key produces an endpoint whose pushes the backend's current VAPID
// signature can't authenticate — the push service 403s it and it becomes a
// silently-dead row — so reusing it must be avoided. `options` is unset in a
// few older engines: treat "can't tell" as a match rather than churn a
// perfectly good subscription.
function subscriptionMatchesVapidKey(subscription, vapidKey) {
  try {
    const current = subscription.options?.applicationServerKey;
    if (!current) return true;
    const a = new Uint8Array(current);
    const b = urlBase64ToUint8Array(vapidKey);
    return a.length === b.length && a.every((byte, i) => byte === b[i]);
  } catch {
    return true;
  }
}

// The reminder-time vs. reminder-interval rows are mutually exclusive
// (see reminder_mode) and both only ever meaningful while the daily
// reminder itself is on — split out from renderToggles() so the
// mode-select and toggle change handlers can re-run just this bit without
// re-touching every other control's value on every small change.
function renderReminderTimingRows() {
  const enabled = preferences.daily_reminder_enabled;
  const isInterval = preferences.reminder_mode === "interval";
  el("reminder-mode-row").hidden = !enabled;
  el("reminder-time-row").hidden = !enabled || isInterval;
  el("reminder-interval-row").hidden = !enabled || !isInterval;
}

function renderToggles() {
  el("push-master-toggle").checked = preferences.push_enabled;
  el("notification-options").hidden = !preferences.push_enabled;
  el("reminder-toggle").checked = preferences.daily_reminder_enabled;
  el("reminder-mode-select").value = preferences.reminder_mode;
  el("reminder-time").value = preferences.daily_reminder_time;
  el("reminder-interval-select").value = String(preferences.reminder_interval_hours);
  renderReminderTimingRows();
  el("smart-nudge-toggle").checked = preferences.smart_nudges_enabled;
  el("weekly-recap-toggle").checked = preferences.weekly_recap_enabled;
  el("quiet-hours-start").value = preferences.quiet_hours_start;
  el("quiet-hours-end").value = preferences.quiet_hours_end;
}

// Debounced so N preference changes made in quick succession (flipping a
// couple of toggles, adjusting both quiet-hours fields back to back) collapse
// into ONE PUT instead of N round trips — each one was previously firing
// immediately on its own "change" event. Optimistic apply is still instant
// (renderToggles() reflects every change the moment it's made, same as
// before); only the actual network write is delayed and batched.
const SAVE_DEBOUNCE_MS = 500;
// How long the quiet "Saved" confirmation stays visible before fading —
// replaces the old per-toggle success toast (push/smart-nudge/weekly-recap
// all used to show one), which fired on top of whatever the user was still
// actively doing in the sheet. A brief ambient confirmation instead of a
// modal interruption.
const SAVED_INDICATOR_VISIBLE_MS = 2000;

let saveTimer = null;
let savedIndicatorTimer = null;
// Snapshot of `preferences` from just BEFORE the current pending batch's
// first change — captured once per batch (null between batches), not
// per-patch, so a failed flush rolls back the WHOLE batch atomically rather
// than leaving it partially applied.
let pendingBatchSnapshot = null;

function showSavedIndicator() {
  const indicator = el("reminders-save-indicator");
  if (!indicator) return;
  indicator.textContent = t("reminders.savedIndicator");
  indicator.classList.add("visible");
  clearTimeout(savedIndicatorTimer);
  savedIndicatorTimer = setTimeout(() => indicator.classList.remove("visible"), SAVED_INDICATOR_VISIBLE_MS);
}

async function flushSave() {
  saveTimer = null;
  const toSend = preferences;
  const rollbackTo = pendingBatchSnapshot;
  pendingBatchSnapshot = null;
  try {
    await api.updateNotificationPreferences(toSend);
    showSavedIndicator();
  } catch {
    // Roll back to what the backend actually has — without this, a failed
    // save (flaky connection, backend hiccup) left the UI showing e.g.
    // "hourly reminders" selected while the server silently kept whatever
    // config it had before, so the schedule the user sees on screen and the
    // one actually running server-side would quietly disagree. The backend
    // is the sole source of truth for what fires (services/
    // notification_scheduler.py), so the UI must reflect its real state,
    // not the un-persisted optimistic guess. Still a toast, deliberately —
    // this is a real failure the user should notice, unlike the quiet
    // success case above.
    preferences = rollbackTo;
    renderToggles();
    showToast(t("reminders.saveFailedToast"), "error");
  }
}

// `immediate` skips the debounce and flushes right away — used only for
// push_enabled (see the master toggle handler below): that change is always
// paired with a real, already-awaited subscribe/unsubscribe side effect, so
// leaving the matching preference row un-persisted for a debounce window is
// a real risk (navigate away/close the tab inside that window and the
// backend never learns push is on, even though a live subscription now
// exists) — not just a missed batching opportunity, unlike every other
// field here.
function savePreferences(patch, { immediate = false } = {}) {
  // Stamps the CURRENT UI language onto every save, not just an explicit
  // language-change event — see backend/services/notification_scheduler.py's
  // localized copy, which reads this stored value for every push it sends
  // (a background sweep has no live request to read a language header
  // from). This keeps it correct with zero extra plumbing even for a user
  // who switches language once and never touches Settings → Notifications
  // again afterward.
  if (pendingBatchSnapshot === null) pendingBatchSnapshot = preferences;
  preferences = { ...preferences, ...patch, language: getLanguage() }; // optimistic, applied immediately regardless of debounce
  clearTimeout(saveTimer);
  if (immediate) {
    flushSave();
  } else {
    saveTimer = setTimeout(flushSave, SAVE_DEBOUNCE_MS);
  }
}

async function subscribeAndRegister() {
  const registration = await navigator.serviceWorker.ready;
  let subscription = await registration.pushManager.getSubscription();
  // Strictly reuse-or-replace: an existing subscription is kept ONLY if it
  // was created with the VAPID key we still use. A mismatch (key rotated, or
  // a leftover from another deploy) is torn down first so the subscribe
  // below mints one fresh endpoint the backend can actually authenticate —
  // never two.
  if (subscription && !subscriptionMatchesVapidKey(subscription, VAPID_PUBLIC_KEY)) {
    await subscription.unsubscribe().catch(() => {});
    subscription = null;
  }
  if (!subscription) {
    subscription = await registration.pushManager.subscribe({
      userVisibleOnly: true,
      applicationServerKey: urlBase64ToUint8Array(VAPID_PUBLIC_KEY),
    });
  }
  await api.subscribePush(subscription.toJSON(), getDeviceId());
}

// Full teardown, not just flipping the preference flag — see
// sql/schema.sql's push_enabled column comment for why: a subscription this
// user has switched off would otherwise sit forever un-pinged (and
// therefore never hit the backend's own 410-Gone self-cleaning path), which
// is exactly the kind of dead-row bloat this whole feature is supposed to
// avoid. Re-enabling later is still cheap (no fresh permission prompt).
async function unsubscribeAndDeregister() {
  if (!("serviceWorker" in navigator)) return;
  const registration = await navigator.serviceWorker.getRegistration();
  const subscription = await registration?.pushManager.getSubscription();
  if (!subscription) return;
  const endpoint = subscription.endpoint;
  await subscription.unsubscribe();
  await api.unsubscribePush(endpoint).catch(() => {
    /* best-effort — a now-invalid endpoint will also self-clean on its next attempted send */
  });
}

async function enablePush() {
  if (!VAPID_PUBLIC_KEY || !pushSupported()) {
    el("push-unavailable-hint").hidden = false;
    return false;
  }
  const permission = await Notification.requestPermission();
  if (permission !== "granted") {
    showToast(t("reminders.permissionDenied"), "error");
    return false;
  }
  try {
    await subscribeAndRegister();
    return true;
  } catch {
    showToast(t("reminders.subscribeFailedToast"), "error");
    return false;
  }
}

// Covers a returning session where permission + a live browser subscription
// already exist and preferences say push is on: makes sure the backend
// still has the CURRENT subscription on file, since the browser can rotate
// it on its own between visits (see sw.js's pushsubscriptionchange handler
// for the other half of this — that one covers the case where the rotation
// happens while this tab isn't even open).
async function resyncExistingSubscription() {
  if (!pushSupported() || !VAPID_PUBLIC_KEY) return;
  if (Notification.permission !== "granted") return;
  try {
    const registration = await navigator.serviceWorker.ready;
    const subscription = await registration.pushManager.getSubscription();
    // Keyed by device_id server-side, so this is self-healing: if the
    // browser rotated the endpoint while no tab was open, this POST both
    // registers the new endpoint AND drops the stale row for this device.
    if (subscription) await api.subscribePush(subscription.toJSON(), getDeviceId());
  } catch {
    /* best-effort resync — a real reminder firing later surfaces any persistent problem */
  }
}

export function initNotifications() {
  const masterToggle = el("push-master-toggle");
  if (!masterToggle) return; // settings markup not present (shouldn't happen in the real app shell)

  if (!pushSupported()) {
    el("push-unavailable-hint").hidden = false;
    masterToggle.disabled = true;
  }

  api
    .getNotificationPreferences()
    .then((prefs) => {
      preferences = { ...DEFAULT_PREFERENCES, ...prefs };
      preferencesLoaded = true;
      renderToggles();
      if (preferences.push_enabled) resyncExistingSubscription();
    })
    .catch(() => {
      /* Settings just show defaults on a fetch failure — not fatal, no toast on first load */
    });

  masterToggle.addEventListener("change", async (event) => {
    const wantsOn = event.target.checked;
    if (wantsOn) {
      const ok = await enablePush();
      if (!ok) {
        masterToggle.checked = false;
        return;
      }
      savePreferences({ push_enabled: true }, { immediate: true });
      renderToggles();
    } else {
      await unsubscribeAndDeregister();
      savePreferences({ push_enabled: false }, { immediate: true });
      renderToggles();
    }
  });

  el("reminder-toggle").addEventListener("change", (event) => {
    savePreferences({ daily_reminder_enabled: event.target.checked });
    renderReminderTimingRows();
  });

  el("reminder-mode-select").addEventListener("change", (event) => {
    savePreferences({ reminder_mode: event.target.value });
    renderReminderTimingRows();
  });

  el("reminder-time").addEventListener("change", (event) => {
    savePreferences({ daily_reminder_time: event.target.value });
  });

  el("reminder-interval-select").addEventListener("change", (event) => {
    savePreferences({ reminder_interval_hours: Number(event.target.value) });
  });

  el("smart-nudge-toggle").addEventListener("change", (event) => {
    savePreferences({ smart_nudges_enabled: event.target.checked });
  });

  el("weekly-recap-toggle").addEventListener("change", (event) => {
    savePreferences({ weekly_recap_enabled: event.target.checked });
  });

  el("quiet-hours-start").addEventListener("change", (event) => {
    savePreferences({ quiet_hours_start: event.target.value });
  });
  el("quiet-hours-end").addEventListener("change", (event) => {
    savePreferences({ quiet_hours_end: event.target.value });
  });

  el("test-notification-btn").addEventListener("click", async () => {
    try {
      await api.sendTestNotification();
      showToast(t("reminders.testSentToast"), "success");
    } catch {
      showToast(t("reminders.testFailedToast"), "error");
    }
  });

  // sw.js's pushsubscriptionchange handler hands a freshly browser-rotated
  // subscription to every open client via postMessage (it has no
  // authenticated Supabase session available to call the backend directly
  // from inside a service worker) — this is the other half of that handoff.
  if ("serviceWorker" in navigator) {
    navigator.serviceWorker.addEventListener("message", (event) => {
      if (event.data?.type !== "ironlog:push-subscription-changed") return;
      // Register the rotated endpoint (device_id-keyed, so it REPLACES this
      // device's row, not adds one) and — belt and braces for a push
      // service that keeps the old endpoint briefly deliverable — explicitly
      // delete the pre-rotation endpoint too.
      api.subscribePush(event.data.subscription, getDeviceId()).catch(() => {});
      if (event.data.oldEndpoint) api.unsubscribePush(event.data.oldEndpoint).catch(() => {});
    });
  }

  // A language switch (Settings' own language toggle) should retarget which
  // language future background pushes arrive in too, not just the visible
  // UI — savePreferences() already stamps getLanguage() onto every save, so
  // an empty patch here is enough to push just that one field. Guarded on
  // preferencesLoaded so this can never fire (and overwrite real settings
  // with placeholder defaults) before the initial GET above has resolved.
  onLanguageChange(() => {
    if (preferencesLoaded) savePreferences({});
  });
}
